File: algorithms/0260.single_number_iii/single_number_iii.rs
use std::collections::HashMap;
impl Solution {
    pub fn single_number(nums: Vec<i32>) -> Vec<i32> {
        let mut diff = nums.iter().fold(0i32, |r, n| r ^ n);
        diff &= -diff;
        let mut unique_nums = vec![0, 0];
        for &n in nums.iter() {
            if n & diff != 0 {
                unique_nums[0] ^= n;
            } else {
                unique_nums[1] ^= n;
            }
        }
        unique_nums
    }
}

File: algorithms/0260.single_number_iii/single_number_iii.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int a = 0, b = 0;
        int64_t t = 0;
        for (const int &n: nums) {
            t ^= n;
        }
        t = t & (-t);
        for (const int &n: nums) {
            if ((n & t) == 0) {
                a ^= n;
            } else {
                b ^= n;
            }
        }
        return {a, b};
    }
};